core/multivm_deps/vm_virtual_blocks/src/old_vm/oracles/storage.rs

```diff
@@ -144,16 +169,26 @@ impl<S: WriteStorage, H: HistoryMode> StorageOracle<S, H> {
             .borrow_mut()
             .is_write_initial(&storage_key);
 
-        get_pubdata_price_bytes(query, is_initial_write)
+        let initial_value = self
+            .get_initial_value(&storage_key)
+            .unwrap_or(query.read_value);
+
+        get_pubdata_price_bytes(initial_value, query.written_value, is_initial_write)
     }
 
     // Returns the price of the update in terms of pubdata bytes.
     // TODO (SMA-1701): update VM to accept gas instead of pubdata.
-    fn value_update_price(&self, query: &LogQuery) -> u32 {
+    fn value_update_price(&mut self, query: &LogQuery) -> u32 {
         let storage_key = storage_key_of_log(query);
 
         let base_cost = self.base_price_for_write(query);
 
+        let initial_value = self
+            .get_initial_value(&storage_key)
+            .unwrap_or(query.read_value);
```

Comment:
If the initial value is not present, you should read it from storage. In `zk_evm` the `read_value` provided is always zero. The correct `read_value` is actually set in the `vm` crate: https://github.com/matter-labs/zksync-era/blob/040374900656c854a7b9de32e5dbaf47c1c47889/core/lib/vm/src/oracles/storage.rs#L121

When the `estimate_refunds_for_write` function is called, the `read_value` is not inserted anywhere